### PR TITLE
Change file system copy folder to copy directory

### PR DESCRIPTION
**Pull request by xster**
Take into account source and destination's file systems so cross-filesystem copies can be made. 

Added copy tests

### CODE REVIEW

**Comment by abarth** [comment]
@tvolkert

File: packages/flutter_tools/lib/src/base/file_system.dart [hunk 1]
```diff
@@ -29,23 +29,25 @@ void ensureDirectoryExists(String filePath) {
...
     String newPath = path.join(destDir.path, path.basename(entity.path));
     if (entity is File) {
-      File newFile = fs.file(newPath);
+      File newFile = destDir.fileSystem.file(newPath);
+      newFile.createSync();
       newFile.writeAsBytesSync(entity.readAsBytesSync());
     } else if (entity is Directory) {
-      copyFolderSync(entity.path, newPath);
+      copyDirectorySync(
+        srcDir.fileSystem.directory(entity.path), destDir.fileSystem.directory(newPath));
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/base/file_system.dart]: s/srcDir.fileSystem.directory(entity.path)/entity/

File: packages/flutter_tools/lib/src/base/file_system.dart [hunk 2]
```diff
@@ -29,23 +29,25 @@ void ensureDirectoryExists(String filePath) {
...
-  Directory destDir = fs.directory(destPath);
   if (!destDir.existsSync())
     destDir.createSync(recursive: true);
 
   srcDir.listSync().forEach((FileSystemEntity entity) {
     String newPath = path.join(destDir.path, path.basename(entity.path));
     if (entity is File) {
-      File newFile = fs.file(newPath);
+      File newFile = destDir.fileSystem.file(newPath);
+      newFile.createSync();
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/base/file_system.dart]: Not needed - `writeAsBytesSync()` will create it on-demand.

File: packages/flutter_tools/test/src/base/file_system_test.dart [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import 'package:file/memory.dart';
```
Comment (tvolkert) [inline packages/flutter_tools/test/src/base/file_system_test.dart]: Needs copyright header

File: packages/flutter_tools/test/src/base/file_system_test.dart [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import 'package:flutter_tools/src/base/file_system.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group('file_system', () {
```
Comment (tvolkert) [inline packages/flutter_tools/test/src/base/file_system_test.dart]: You could instantiate a second `MemoryFileSystem` for the destination directory and make sure there's no cross-pollution between the two file systems.

File: packages/flutter_tools/test/src/base/file_system_test.dart [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    /// Test file_systems.copyDirectorySync() using MemoryFileSystem.
+    test('test directory copy', () async {
+      MemoryFileSystem memoryFs = new MemoryFileSystem();
+      String sourcePath = '/some/origin';
+      Directory sourceDirectory = await memoryFs.directory(sourcePath).create(recursive: true);
+      memoryFs.currentDirectory = sourcePath;
+      (await memoryFs.file('some_file.txt').create()).writeAsStringSync('bleh');
```
Comment (tvolkert) [inline packages/flutter_tools/test/src/base/file_system_test.dart]: No need to create - `writeAs*` will create it on demand.

File: packages/flutter_tools/lib/src/base/file_system.dart (hunk 1 above)
Comment (xster) [inline packages/flutter_tools/lib/src/base/file_system.dart]: Doh. I weaksauced

File: packages/flutter_tools/lib/src/base/file_system.dart (hunk 2 above)
Comment (xster) [inline packages/flutter_tools/lib/src/base/file_system.dart]: Done

File: packages/flutter_tools/test/src/base/file_system_test.dart (hunk 3 above)
Comment (xster) [inline packages/flutter_tools/test/src/base/file_system_test.dart]: Done

File: packages/flutter_tools/test/src/base/file_system_test.dart (hunk 4 above)
Comment (xster) [inline packages/flutter_tools/test/src/base/file_system_test.dart]: Done. PTAL.

File: packages/flutter_tools/test/src/base/file_system_test.dart (hunk 5 above)
Comment (xster) [inline packages/flutter_tools/test/src/base/file_system_test.dart]: Done

**Comment by tvolkert** [comment]
LGTM!

**Pull request merged**